

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>An introduction to DMD &mdash; flowTorch 0.1 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="SVD of the flow past a circular cylinder" href="svd_cylinder.html" />
    <link rel="prev" title="PSP data explorer and loader" href="ipsp_explorer.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> flowTorch
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Examples</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../overview/glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="linear_algebra_basics.html">Linear algebra with PyTorch</a></li>
<li class="toctree-l1"><a class="reference internal" href="svd_ht.html">Optimal hard threshold for singular values</a></li>
<li class="toctree-l1"><a class="reference internal" href="ipsp_explorer.html">PSP data explorer and loader</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">An introduction to DMD</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#DMD-at-a-glance">DMD at a glance</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Goal">Goal</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Inputs">Inputs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Basic-idea-and-theory">Basic idea and theory</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Steps-to-compute-the-DMD-in-practice">Steps to compute the DMD in practice</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#A-simple-example">A simple example</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Generating-the-dataset">Generating the dataset</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Singular-Value-Decomposition-(SVD)">Singular Value Decomposition (SVD)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Truncating">Truncating</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Eigen-decomposition">Eigen decomposition</a></li>
<li class="toctree-l3"><a class="reference internal" href="#DMD-modes">DMD modes</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="svd_cylinder.html">SVD of the flow past a circular cylinder</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmd_cylinder.html">DMD of the flow past a circular cylinder</a></li>
<li class="toctree-l1"><a class="reference internal" href="cnm_cylinder.html">CNM of the flow past a circular cylinder</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmd_naca0012_surface.html">SVD and DMD analysis of airfoil surface data</a></li>
<li class="toctree-l1"><a class="reference internal" href="piv_analysis.html">Analyzing PIV data of a longitudinal vortex</a></li>
</ul>
<p class="caption"><span class="caption-text">Python API</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../flowtorch.html">flowtorch</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">flowTorch</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>An introduction to DMD</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/notebooks/dmd_intro.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<p><img alt="CC" src="https://i.creativecommons.org/l/by/4.0/88x31.png" /></p>
<p>This work is licensed under a <a class="reference external" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.</p>
<div class="section" id="An-introduction-to-DMD">
<h1>An introduction to DMD<a class="headerlink" href="#An-introduction-to-DMD" title="Permalink to this headline">¶</a></h1>
<p>This notebook introduces the basic theory behind dynamic mode decomposition (DMD). The content is heavily influenced by the following publications, which are also great resources for learning:</p>
<ul class="simple">
<li><p><a class="reference external" href="http://www.pyrunner.com/weblog/2016/07/25/dmd-python/">Dynamic Mode Decomposition in Python</a> by Robert Taylor</p></li>
<li><p><a class="reference external" href="https://www.cambridge.org/core/books/datadriven-science-and-engineering/77D52B171B60A496EAFE4DB662ADC36E">Data Driven Science and Engineering</a> by Steven L. Brunton and J. Nathan Kutz</p></li>
<li><p><a class="reference external" href="https://epubs.siam.org/doi/abs/10.1137/1.9781611974508">Dynamic Mode Decomposition</a> by J. Nathan Kutz, Steven L. Brunton, Bingni W. Brunton, and Joshua L. Proctor</p></li>
<li><p><a class="reference external" href="https://raw.githubusercontent.com/mathLab/PyDMD/master/tutorials/tutorial_dsweb.pdf">PyDMD tutorial</a> by Niccola Demo and Marco Tezzele</p></li>
</ul>
<div class="section" id="DMD-at-a-glance">
<h2>DMD at a glance<a class="headerlink" href="#DMD-at-a-glance" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Goal">
<h3>Goal<a class="headerlink" href="#Goal" title="Permalink to this headline">¶</a></h3>
<p>There are three typical goals when performing a DMD analysis:</p>
<ul class="simple">
<li><p>modal analysis; in the context of fluid mechanics, modal analysis may be seen as the extraction of coherent flow structures and their associated frequencies and amplitudes</p></li>
<li><p>creation of a reduced-order model (ROM)</p></li>
<li><p>stability analysis; under certain conditions, particular modes might be identified as growing or shrinking over time</p></li>
</ul>
</div>
<div class="section" id="Inputs">
<h3>Inputs<a class="headerlink" href="#Inputs" title="Permalink to this headline">¶</a></h3>
<p>DMD is a data-driven method that requires sequential snapshot data as input. Applied to fluid flows, snapshots of the pressure or velocity field or derived quantitites are typically analyzed with DMD. Suppose we take a snapshot at time instance <span class="math notranslate nohighlight">\(t_n\)</span> and organize it in a long column vector <span class="math notranslate nohighlight">\(\mathbf{x}_n\)</span>. The spatial arangement of points in the snapshot is ignored (of course, all column vectors must be assembled in the same order). We can then organize a time series of <span class="math notranslate nohighlight">\(N\)</span>
snapshots in two distinct data matrices:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\mathbf{X} =
 \begin{bmatrix}
   \vert &amp; \vert &amp;     &amp; \vert \\
   \mathbf{x}_0  &amp; \mathbf{x}_1   &amp; ... &amp; \mathbf{x}_{N-2} \\
   \vert &amp; \vert &amp;     &amp; \vert
 \end{bmatrix}\end{split}\]</div>
<div class="math notranslate nohighlight">
\[\begin{split}\mathbf{X}^\prime =
 \begin{bmatrix}
   \vert &amp; \vert &amp;     &amp; \vert \\
   \mathbf{x}_1  &amp; \mathbf{x}_2   &amp; ... &amp; \mathbf{x}_{N-1} \\
   \vert &amp; \vert &amp;     &amp; \vert
 \end{bmatrix}\end{split}\]</div>
<p>The second data matrix is similar to the first one but shifted by one time instance. The reason for this arrangement will become clear in the next section.</p>
</div>
<div class="section" id="Basic-idea-and-theory">
<h3>Basic idea and theory<a class="headerlink" href="#Basic-idea-and-theory" title="Permalink to this headline">¶</a></h3>
<p>The DMD’s main objective is to approximate a linear operator <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> that maps a snapshot from time <span class="math notranslate nohighlight">\(t_n\)</span> to <span class="math notranslate nohighlight">\(t_{n+1}\)</span> as:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}_{n+1} = \mathbf{Ax}_n\]</div>
<p>Of course, if the underlying dynamics are not perfectly linear, there won’t be a single linear operator that works for all snapshot pairs, but we can approximate a best-fit linear operator using the pseudo inverse <span class="math notranslate nohighlight">\(\mathbf{X}^\dagger\)</span>:</p>
<div class="math notranslate nohighlight">
\[\mathbf{A}\triangleq \mathbf{X}^\prime\mathbf{X}^\dagger\]</div>
<p>Knowing the linear operator <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> allows us to make predictions beyond the time range of the snapshots by repeatedly mapping <span class="math notranslate nohighlight">\(\mathbf{x}_n\)</span> to <span class="math notranslate nohighlight">\(\mathbf{x}_{n+1}\)</span>. If we also know the eigen decomposition of <span class="math notranslate nohighlight">\(\mathbf{A}\)</span>, namely:</p>
<div class="math notranslate nohighlight">
\[\mathbf{A} = \mathbf{\Phi\Lambda\Phi}^{-1},\]</div>
<p>we can derive a relatively simple equation to make future state predications at any time <span class="math notranslate nohighlight">\(t_n\)</span> without having to make <span class="math notranslate nohighlight">\(n\)</span> discrete steps to get there. In the equation above, <span class="math notranslate nohighlight">\(\mathbf{\Phi}\)</span> is a matrix whose columns are the eigenvectors of <span class="math notranslate nohighlight">\(\mathbf{A}\)</span>, and <span class="math notranslate nohighlight">\(\mathbf{\Lambda}\)</span> is a diagonal matrix with the eigenvalues of <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> forming the diagonal elements. The discrete operator <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> propagates <span class="math notranslate nohighlight">\(\mathbf{x}_n\)</span> to
<span class="math notranslate nohighlight">\(\mathbf{x}_{n+1}\)</span>. If we wanted to know <span class="math notranslate nohighlight">\(\mathbf{x}_{n+2}\)</span>, we could apply <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> twice, e.g., <span class="math notranslate nohighlight">\(\mathbf{x}_{n+2}=\mathbf{AAx}_{n}\)</span>. If we then replace <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> with its eigen decomposition and use the identity <span class="math notranslate nohighlight">\(\mathbf{\Phi\Phi}^{-1} = \mathbf{\Phi}^{-1}\mathbf{\Phi} = \mathbf{I}\)</span>, we get:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}_{n+2}=\mathbf{AAx}_{n} = \mathbf{\Phi\Lambda\Phi}^{-1}\mathbf{\Phi\Lambda\Phi}^{-1}\mathbf{x}_{n} = \mathbf{\Phi\Lambda}^{2}\mathbf{\Phi}^{-1}\mathbf{x}_{n},\]</div>
<p>or more in general for instance <span class="math notranslate nohighlight">\(n+k\)</span></p>
<div class="math notranslate nohighlight">
\[\mathbf{x}_{n+k} = \mathbf{\Phi\Lambda}^{k}\mathbf{\Phi}^{-1}\mathbf{x}_{n}.\]</div>
<p>Now we use the expression above to re-write the equation <span class="math notranslate nohighlight">\(\mathbf{X}^\prime \approx \mathbf{AX}\)</span> with the eigen decomposition of <span class="math notranslate nohighlight">\(\mathbf{A}\)</span>:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\begin{align}
  \mathbf{X}^\prime &amp;= \mathbf{AX} \\
  &amp;= \mathbf{A}\left[\mathbf{x}_0, \mathbf{x}_1, ... , \mathbf{x}_{N-2} \right]\\
  &amp;= \left[\mathbf{\Phi\Lambda}^{0}\mathbf{\Phi}^{-1}\mathbf{x}_0, \mathbf{\Phi\Lambda}^{1}\mathbf{\Phi}^{-1}\mathbf{x}_0, ... , \mathbf{\Phi\Lambda}^{N-2}\mathbf{\Phi}^{-1}\mathbf{x}_0 \right]\\
  &amp;= \mathbf{\Phi}\left[\mathbf{\Lambda}^{0}, \mathbf{\Lambda}^{1}, ... , \mathbf{\Lambda}^{N-2}\right]\mathbf{\Phi}^{-1}\mathbf{x}_0.
\end{align}\end{split}\]</div>
<p>The initial state <span class="math notranslate nohighlight">\(\mathbf{x}_0\)</span> projected onto <span class="math notranslate nohighlight">\(\mathbf{\Phi}^{-1}\)</span> is often abbreviated as <span class="math notranslate nohighlight">\(\mathbf{b}=\mathbf{\Phi}^{-1}\mathbf{x}_0\)</span>. Then, the generic formula to predict the state vector <span class="math notranslate nohighlight">\(\mathbf{x}_n\)</span> reads:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}_n = \mathbf{\Phi\Lambda}^{n}\mathbf{b}.\]</div>
<p>The same problem may be posed with a time-continuous state <span class="math notranslate nohighlight">\(\mathbf{x}=\mathbf{x}(t)\)</span> instead of discrete one. Then the basic formulation of DMD, mapping a given state linearly into the future, becomes a system of linear ordinary differential equations (ODEs):</p>
<div class="math notranslate nohighlight">
\[\frac{\mathrm{d}\mathbf{x}}{\mathrm{d}t} = \mathbf{\mathcal{A}x},\]</div>
<p>where <span class="math notranslate nohighlight">\(\mathbf{\mathcal{A}}\)</span> is the continuous counterpart of <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> (the connection between both operators will be established in the following paragraphs). The initial value problem has the solution:</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}(t) = \mathrm{exp}\left(\mathbf{\mathcal{A}}t\right) \mathbf{x}_0,\]</div>
<p>where <span class="math notranslate nohighlight">\(\mathrm{exp}\left(\mathbf{\mathcal{A}}\right)\)</span> is the so-called <a class="reference external" href="https://en.wikipedia.org/wiki/Matrix_exponential">matrix exponential</a> (refer to chapter 7.7 of <a class="reference external" href="https://www.wiley.com/en-us/Elementary+Differential+Equations+and+Boundary+Value+Problems%2C+11th+Edition-p-9781119381648">this book</a> for the derivation). Why do we need the continuous formulation you may wonder. The answer is that it is much easier to assess periodicity and stability in this formulation. Say we know
the eigen decomposition <span class="math notranslate nohighlight">\(\mathbf{\mathcal{A}} = \tilde{\mathbf{\Phi}}\tilde{\mathbf{\Lambda}}\tilde{\mathbf{\Phi}}^{-1}\)</span>, then we can reformulate the analytical solution as (refer to the article on matrix exponentials):</p>
<div class="math notranslate nohighlight">
\[\mathbf{x}(t) = \mathrm{exp}\left(\tilde{\mathbf{\Phi}}\tilde{\mathbf{\Lambda}}\tilde{\mathbf{\Phi}}^{-1}t\right) \mathbf{x}_0 = \tilde{\mathbf{\Phi}}\mathrm{exp}\left(\tilde{\mathbf{\Lambda}}t\right)\tilde{\mathbf{\Phi}}^{-1} \mathbf{x}_0 = \tilde{\mathbf{\Phi}}\mathrm{exp}\left(\tilde{\mathbf{\Lambda}}t\right)\tilde{\mathbf{b}}.\]</div>
<p>Each column of <span class="math notranslate nohighlight">\(\tilde{\mathbf{\Phi}}\)</span> has the same dimension as a snapshot and may be interpreted as a characteristic (spatial) flow structure. The exponential term yields a diagonal matrix with time-dependent coefficients <span class="math notranslate nohighlight">\(\tilde{\lambda}_i t\)</span>, where <span class="math notranslate nohighlight">\(\tilde{\lambda}_i\)</span> is the <span class="math notranslate nohighlight">\(i\)</span>th eigenvalue of <span class="math notranslate nohighlight">\(\mathbf{\mathcal{A}}\)</span>. If we multiply the eigenvector matrix <span class="math notranslate nohighlight">\(\tilde{\mathbf{\Phi}}\)</span> with the matrix exponential, we are effectively multiplying the
<span class="math notranslate nohighlight">\(i\)</span>th column of <span class="math notranslate nohighlight">\(\tilde{\mathbf{\Phi}}\)</span> with <span class="math notranslate nohighlight">\(\mathrm{exp}(\tilde{\lambda}_i t)\)</span>. Therefore, the eigenvalue <span class="math notranslate nohighlight">\(\tilde{\lambda}_i\)</span> defines the <span class="math notranslate nohighlight">\(i\)</span>th mode’s temporal behavior. Specifically, we may use Euler’s identity to write:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\begin{align}
  \tilde{\lambda}_j &amp;= a_i + b_i j \\
  \mathrm{exp}\left(\tilde{\lambda}_i t\right) &amp;= \mathrm{exp}\left((a_i+b_i j) t\right)\\
  &amp;= \mathrm{exp}\left(a_i t\right)\mathrm{exp}\left(b_i j t\right) \\
  &amp;= \mathrm{exp}\left(a_i t\right) \left(\mathrm{cos}\left(b_i t\right)+j\mathrm{sin}\left(b_i t\right)\right),
\end{align}\end{split}\]</div>
<p>where <span class="math notranslate nohighlight">\(a_i = Re(\tilde{\lambda}_i)\)</span> and <span class="math notranslate nohighlight">\(b_i = Im(\tilde{\lambda}_i)\)</span> are the real and imaginary parts of the <span class="math notranslate nohighlight">\(i\)</span>th eigenvalue, respectively. Now we can observe the following:</p>
<ul class="simple">
<li><p>if the eigenvalue has a non-zero imaginary part <span class="math notranslate nohighlight">\(b_i\)</span>, the time behavior will be periodic</p></li>
<li><p>if we express the argument of sine and cosine as <span class="math notranslate nohighlight">\(b_i t = 2\pi f_i t\)</span>, the formula for the <span class="math notranslate nohighlight">\(i\)</span>th frequency is <span class="math notranslate nohighlight">\(f_i = b_i/(2\pi)\)</span></p></li>
<li><p>for the real part <span class="math notranslate nohighlight">\(a_i\)</span></p>
<ul>
<li><p>if <span class="math notranslate nohighlight">\(a_i = 0\)</span>, the temporal behavior is constant or periodic with constant amplitude</p></li>
<li><p>if <span class="math notranslate nohighlight">\(a_i &gt; 0\)</span>, the mode is growing over time</p></li>
<li><p>if <span class="math notranslate nohighlight">\(a_i &lt; 0\)</span>, the mode is decreasing over time</p></li>
</ul>
</li>
<li><p>because of the points above, <span class="math notranslate nohighlight">\(a_i\)</span> may be called the growth rate of the <span class="math notranslate nohighlight">\(i\)</span>th mode</p></li>
</ul>
<p>Typically, we do not know <span class="math notranslate nohighlight">\(\mathbf{\mathcal{A}}\)</span> but we have to work with discrete snapshots and the operator <span class="math notranslate nohighlight">\(\mathbf{A}\)</span>. Therefore, we need a link between the discrete and the continuous formulation, which is established by evaluating the continuous solution at times <span class="math notranslate nohighlight">\(t_n=n\Delta t\)</span> and <span class="math notranslate nohighlight">\(t_{n+1}=(n+1)\Delta t\)</span> to obtain <span class="math notranslate nohighlight">\(\mathbf{x}_{n}\)</span> and <span class="math notranslate nohighlight">\(\mathbf{x}_{n+1}\)</span>, respectively:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\begin{align}
  \mathbf{x}(t=n\Delta t) &amp;= \mathrm{exp}\left(\mathbf{\mathcal{A}}n\Delta t\right)\mathbf{x}_0 \\
  \mathbf{x}(t=(n+1)\Delta t) &amp;= \mathrm{exp}\left(\mathbf{\mathcal{A}}(n+1)\Delta t\right)\mathbf{x}_0 \\
  &amp;= \mathrm{exp}\left(\mathbf{\mathcal{A}}\Delta t\right) \mathrm{exp}\left(\mathbf{\mathcal{A}}n \Delta t\right) \mathbf{x}_0\\
  \mathbf{x}_{n+1} &amp;= \underbrace{\mathrm{exp}\left(\mathbf{\mathcal{A}}\Delta t\right)}_{\mathbf{A}}\mathbf{x}_n \\
  \mathbf{x}_{n+1} &amp;= \mathbf{Ax}_n
\end{align}\end{split}\]</div>
<p>By comparing the eigen decomposition of <span class="math notranslate nohighlight">\(\mathbf{\mathcal{A}}\)</span> and <span class="math notranslate nohighlight">\(\mathbf{A}\)</span>, we notice that the eigenvectors are identical, i.e., <span class="math notranslate nohighlight">\(\mathbf{\Phi} = \tilde{\mathbf{\Phi}}\)</span>. The eigenvalues are related by the relation:</p>
<div class="math notranslate nohighlight">
\[\mathbf{\Lambda} = \mathrm{exp}\left(\tilde{\mathbf{\Lambda}}\Delta t\right).\]</div>
<p>Summarizing the paragraphs above, we have learned:</p>
<ul class="simple">
<li><p>the linear operator <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> maps a given state by <span class="math notranslate nohighlight">\(\Delta t\)</span> into the future</p></li>
<li><p>the operator’s eigenvectors (DMD modes) may represent characteristic spatial flow patterns</p></li>
<li><p>the operator’s eigenvalues define the modes’ temporal behavior in terms of growth and periodicity</p></li>
<li><p>the <span class="math notranslate nohighlight">\(\mathbf{b}\)</span> vector represents a constant weight for each DMD mode; therefore, we may consider the coefficients of <span class="math notranslate nohighlight">\(\mathbf{b}\)</span> as mode amplitudes</p></li>
</ul>
<p>For fluid mechanical problems in particular, there is a pratical issue with the theory described above. If a snapshot consists of <span class="math notranslate nohighlight">\(M\)</span> points, then <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> is an <span class="math notranslate nohighlight">\(M\times M\)</span> dense matrix. If <span class="math notranslate nohighlight">\(M=10^4\)</span>, the resulting operator would require about 0.75GB of memory (double precision). Increasing the number of points to <span class="math notranslate nohighlight">\(M=10^5\)</span> yields a 74.5GB operator, which is already impractical for most applications. The next section details some numerical tricks to overcome this
problem.</p>
</div>
<div class="section" id="Steps-to-compute-the-DMD-in-practice">
<h3>Steps to compute the DMD in practice<a class="headerlink" href="#Steps-to-compute-the-DMD-in-practice" title="Permalink to this headline">¶</a></h3>
<p><strong>1. Compute the SVD of :math:`mathbf{X}`</strong></p>
<p>First, we factorize the data matrix <span class="math notranslate nohighlight">\(\mathbf{X}\)</span> into its singular value decomposition (SVD):</p>
<div class="math notranslate nohighlight">
\[\mathbf{X} = \mathbf{U\Sigma V}^{\dagger},\]</div>
<p>where <span class="math notranslate nohighlight">\(\mathbf{U}\)</span> is a <span class="math notranslate nohighlight">\(M\times M\)</span> unitary matrix, <span class="math notranslate nohighlight">\(\mathbf{\Sigma}\)</span> is a <span class="math notranslate nohighlight">\(M\times N\)</span> matrix containing a smaller diagonal matrix of singular values, and <span class="math notranslate nohighlight">\(\mathbf{V}\)</span> is a <span class="math notranslate nohighlight">\(N\times N\)</span> unitary matrix (refer to the <a class="reference internal" href="linear_algebra_basics.html"><span class="doc">Linear algebra with PyTorch</span></a> notebook). What the SVD essentially provides us with is an easily controllable means to perform matrix approximations and a potentially better coordinate system to represent the system’s
dynamics.</p>
<p><strong>2. Compute a reduced linear operator</strong></p>
<p>As mentioned above, we can approximate the discrete linear operator as:</p>
<div class="math notranslate nohighlight">
\[\mathbf{A}\triangleq \mathbf{X}^\prime \mathbf{X}^\dagger = \mathbf{X}^\prime \mathbf{V\Sigma}^{-1}\mathbf{U}^\dagger.\]</div>
<p>However, to reduce the computational effort, we first represent <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> in the span of the first <span class="math notranslate nohighlight">\(r\)</span> left singular vectors <span class="math notranslate nohighlight">\(\mathbf{U}_r\)</span>, where <span class="math notranslate nohighlight">\(0&lt;r\le N\)</span>:</p>
<div class="math notranslate nohighlight">
\[\tilde{\mathbf{A}} = \mathbf{U}_r^\dagger \mathbf{AU}_r,\]</div>
<p>where <span class="math notranslate nohighlight">\(\tilde{\mathbf{A}}\)</span> is an <span class="math notranslate nohighlight">\(r\times r\)</span> square matrix that is <a class="reference external" href="ttps://en.wikipedia.org/wiki/Matrix_similarity">similar</a> to <span class="math notranslate nohighlight">\(\mathbf{A}\)</span>. Note that this projection is only meaningful if <span class="math notranslate nohighlight">\(\mathbf{X}\)</span> and <span class="math notranslate nohighlight">\(\mathbf{X}^\prime\)</span> do not have full rank, e.g., if there exist coherent structures in the flow. Combining the two expressions above and using the rank <span class="math notranslate nohighlight">\(r\)</span> approximation of the SVD yields:</p>
<div class="math notranslate nohighlight">
\[\tilde{\mathbf{A}} = \mathbf{U}_r^\dagger\mathbf{X}^\prime\mathbf{V}_r\mathbf{\Sigma}_r^{-1}\mathbf{U}_r^\dagger\mathbf{U}_r = \mathbf{U}_r^\dagger\mathbf{X}^\prime\mathbf{V}_r\mathbf{\Sigma}_r^{-1}.\]</div>
<p>It is important to mention that <span class="math notranslate nohighlight">\(\mathbf{U}_r^\dagger\mathbf{U}_r\)</span> yields an <span class="math notranslate nohighlight">\(r\times r\)</span> identity matrix, whereas <span class="math notranslate nohighlight">\(\mathbf{U}_r\mathbf{U}_r^\dagger\)</span> in general <strong>does not</strong> yield a <span class="math notranslate nohighlight">\(M\times M\)</span> identity matrix. After completing this second step, the reduced <span class="math notranslate nohighlight">\(r\times r\)</span> operator <span class="math notranslate nohighlight">\(\tilde{\mathbf{A}}\)</span> is much easier to deal with.</p>
<p><strong>3. Compute the eigen decomposition of the reduced linear operator</strong></p>
<p>If <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> is not of full rank and lies in the span of the first <span class="math notranslate nohighlight">\(r\)</span> singular vectors, <span class="math notranslate nohighlight">\(\tilde{\mathbf{A}}\)</span> and <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> are <a class="reference external" href="https://en.wikipedia.org/wiki/Matrix_similarity">similar</a> and, therefore, have the same non-zero eigenvalues. If <span class="math notranslate nohighlight">\(r\ll M\)</span> the following eigenvalue problem is much easier to solve:</p>
<div class="math notranslate nohighlight">
\[\tilde{\mathbf{A}}\mathbf{W} =\mathbf{W\Lambda}\]</div>
<p>The rank reduction may also introduce an error in the computed eigenvalues.</p>
<p><strong>4. Compute the eigenvectors of the original linear operator</strong></p>
<p>The final piece of the puzzle is the computation to obtain the original operator’s eigenvectors <span class="math notranslate nohighlight">\(\mathbf{\Phi}\)</span>. In the <a class="reference external" href="https://hal-polytechnique.archives-ouvertes.fr/file/index/docid/1020654/filename/DMS0022112010001217a.pdf">original article introducing the DMD</a> by Peter Schmid, the eigenvalues of <span class="math notranslate nohighlight">\(\mathbf{A}\)</span> are recovered by computing:</p>
<div class="math notranslate nohighlight">
\[\tilde{\mathbf{\Phi}} = \mathbf{U}_r\mathbf{W}\]</div>
<p>However, if we try to transform the reduced eigenvalue problem into eigenvalue problem of <span class="math notranslate nohighlight">\(\mathbf{A}\)</span>, we find:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\begin{align}
  \tilde{\mathbf{A}}\mathbf{W} &amp;= \mathbf{W\Lambda} \\
  \mathbf{U}_r\underbrace{\mathbf{U}_r^\dagger\mathbf{AU}_r}_{\tilde{\mathbf{A}}}\mathbf{W} &amp;= \mathbf{U}_r\mathbf{W\Lambda} \\
  \underbrace{\mathbf{U}_r\mathbf{U}_r^\dagger}_{\neq \mathbf{I}} \mathbf{A}\tilde{\mathbf{\Phi}} &amp;= \tilde{\mathbf{\Phi}}\mathbf{\Lambda}
\end{align}\end{split}\]</div>
<p>A more widely accepted definition suggested by <a class="reference external" href="https://arxiv.org/pdf/1312.0041.pdf">Jonathan Tu et al.</a> is:</p>
<div class="math notranslate nohighlight">
\[\mathbf{\Phi} = \mathbf{X}^\prime\mathbf{V}_r\mathbf{\Sigma}_r^{-1}\mathbf{W}.\]</div>
<p>Multiplying both sides of the reduced eigenvalue problem with <span class="math notranslate nohighlight">\(\mathbf{X}^\prime\mathbf{V}_r\mathbf{\Sigma}_r^{-1}\)</span> yields:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\begin{align}
  \tilde{\mathbf{A}}\mathbf{W} &amp;= \mathbf{W\Lambda} \\
  \mathbf{X}^\prime\mathbf{V}_r\mathbf{\Sigma}_r^{-1} \underbrace{\mathbf{U}_r^\dagger\mathbf{X}^\prime\mathbf{V}_r\mathbf{\Sigma}_r^{-1}}_{\tilde{\mathbf{A}}} \mathbf{W}&amp;= \mathbf{X}^\prime\mathbf{V}_r\mathbf{\Sigma}_r^{-1}\mathbf{W\Lambda} \\
  \mathbf{A}\mathbf{X}^\prime\mathbf{V}_r\mathbf{\Sigma}_r^{-1}\mathbf{W}&amp;= \mathbf{X}^\prime\mathbf{V}_r\mathbf{\Sigma}_r^{-1}\mathbf{W\Lambda} \\
  \mathbf{A\Phi} &amp;= \mathbf{\Phi\Lambda}
\end{align}\end{split}\]</div>
<p>Therefore, the computation suggested by Jonathan Tu et al. yields the so-called <strong>exact DMD modes</strong>. The modes obtained using the original definition are commonly called <strong>projected DMD modes</strong> because the relation between both eigenvector matrices is <span class="math notranslate nohighlight">\(\tilde{\mathbf{\Phi}} = \mathbf{U}_r\mathbf{U}_r^\dagger\mathbf{\Phi}\)</span> (we obtain <span class="math notranslate nohighlight">\(\tilde{\mathbf{\Phi}}\)</span> by projecting <span class="math notranslate nohighlight">\(\mathbf{\Phi}\)</span> onto the POD modes/left singular vectors of <span class="math notranslate nohighlight">\(\mathbf{X}\)</span>). However, if the last column
of <span class="math notranslate nohighlight">\(\mathbf{X}^\prime\)</span> is in the span of <span class="math notranslate nohighlight">\(\mathbf{X}\)</span>, both definitions yield the same result; refer to remark 5 of the article by J. Tu et al. linked above.</p>
</div>
</div>
<div class="section" id="A-simple-example">
<h2>A simple example<a class="headerlink" href="#A-simple-example" title="Permalink to this headline">¶</a></h2>
<p>In the following, we’ll create a simple dataset and implement a basic version of the DMD from scratch in PyTorch.</p>
<div class="section" id="Generating-the-dataset">
<h3>Generating the dataset<a class="headerlink" href="#Generating-the-dataset" title="Permalink to this headline">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">torch</span> <span class="k">as</span> <span class="nn">pt</span>

<span class="c1"># increase resolution of plots</span>
<span class="n">mpl</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">160</span>
</pre></div>
</div>
</div>
<p>The problem we are examining has two independent variables: the time <span class="math notranslate nohighlight">\(t\)</span> and the spatial coordinate <span class="math notranslate nohighlight">\(x\)</span>. We start by creating a spatio-temporal grid to create our test data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">x</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">50</span><span class="p">)</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">pi</span><span class="p">,</span> <span class="mi">40</span><span class="p">)</span>
<span class="n">X</span><span class="p">,</span> <span class="n">T</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">meshgrid</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now we evaluate several periodic functions on the grid. There are several ways to visualize the temporal evolution of the 1D profiles (one spatial coordinate), of which we explore two in this notebook. In the first variant, we plot the spatial distribution at fixed time intervals as a line plot. To introduce the notion of time, the lines’ opacity decreases as the time increases. Moreover, we allow to plot optionally both real and imaginary part.</p>
<p>The first function comprises two oscillating hyperbolic functions:</p>
<div class="math notranslate nohighlight">
\[f_1 = 10\mathrm{tanh}(x/2)\mathrm{cosh}(x/2)e^{(0.1+2.8j)t}\]</div>
<p>The positive real part in the exponent gives the oscillation a slowly growing amplitude. The second function is an oscillating parabola:</p>
<div class="math notranslate nohighlight">
\[f_2 = (20-x^2/5)e^{(-0.05+2.3j)t}\]</div>
<p>In comparison to <span class="math notranslate nohighlight">\(f_1\)</span>, this function has a <strong>decreasing</strong> amplitude and oscillates at a slightly smaller frequency (smaller imaginary part). The third function is an oscillating line (like a seesaw):</p>
<div class="math notranslate nohighlight">
\[f_3 = x e^{0.6j t}\]</div>
<p>This time, the oscillation’s amplitude remains constant. In comparison to the other two functions, <span class="math notranslate nohighlight">\(f_3\)</span> oscillates at the lowest frequency.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">plot_as_line</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">every</span><span class="p">,</span> <span class="n">show_imag</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
    <span class="n">t_selected</span> <span class="o">=</span> <span class="n">t</span><span class="p">[::</span><span class="n">every</span><span class="p">]</span>
    <span class="n">data_selected</span> <span class="o">=</span> <span class="n">data</span><span class="p">[:,</span> <span class="p">::</span><span class="n">every</span><span class="p">]</span>
    <span class="n">opacity_factor</span> <span class="o">=</span> <span class="mf">0.5</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">t_selected</span><span class="p">)):</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">data</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">real</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;C0&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="mi">1</span><span class="o">+</span><span class="n">i</span><span class="o">*</span><span class="n">opacity_factor</span><span class="p">),</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;$Re(f)$&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">show_imag</span><span class="p">:</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">data</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">imag</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;C1&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="mi">1</span><span class="o">+</span><span class="n">i</span><span class="o">*</span><span class="n">opacity_factor</span><span class="p">),</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;$Im(f)$&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">data</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">real</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;C0&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="mi">1</span><span class="o">+</span><span class="n">i</span><span class="o">*</span><span class="n">opacity_factor</span><span class="p">),</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">show_imag</span><span class="p">:</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">data</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">imag</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;C1&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="mi">1</span><span class="o">+</span><span class="n">i</span><span class="o">*</span><span class="n">opacity_factor</span><span class="p">),</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;$x$&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;time progression indicated by decreasing opacity&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">f1</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="mf">10.0</span><span class="o">*</span><span class="n">pt</span><span class="o">.</span><span class="n">tanh</span><span class="p">(</span><span class="n">X</span><span class="o">/</span><span class="mf">2.0</span><span class="p">)</span><span class="o">/</span><span class="n">pt</span><span class="o">.</span><span class="n">cosh</span><span class="p">(</span><span class="n">X</span><span class="o">/</span><span class="mf">2.0</span><span class="p">),</span> <span class="n">pt</span><span class="o">.</span><span class="n">exp</span><span class="p">((</span><span class="mf">0.1</span><span class="o">+</span><span class="mf">2.8</span><span class="n">j</span><span class="p">)</span><span class="o">*</span><span class="n">T</span><span class="p">))</span>
<span class="n">plot_as_line</span><span class="p">(</span><span class="n">f1</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_dmd_intro_8_0.png" src="../_images/notebooks_dmd_intro_8_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">f2</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="mi">20</span><span class="o">-</span><span class="mf">0.2</span><span class="o">*</span><span class="n">pt</span><span class="o">.</span><span class="n">pow</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="n">pt</span><span class="o">.</span><span class="n">exp</span><span class="p">((</span><span class="o">-</span><span class="mf">0.05</span><span class="o">+</span><span class="mf">2.3</span><span class="n">j</span><span class="p">)</span><span class="o">*</span><span class="n">T</span><span class="p">))</span>
<span class="n">plot_as_line</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_dmd_intro_9_0.png" src="../_images/notebooks_dmd_intro_9_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">f3</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">pt</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="mf">0.6</span><span class="n">j</span><span class="o">*</span><span class="n">T</span><span class="p">))</span>
<span class="n">plot_as_line</span><span class="p">(</span><span class="n">f3</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_dmd_intro_10_0.png" src="../_images/notebooks_dmd_intro_10_0.png" />
</div>
</div>
<p>The following plot shows a superposition of all three functions, which comprises our DMD dataset. Even though simple functions were used to create the dataset, it is not straight forward any more to infer the individual contributions visually. Our goal is to isolate the individual frequencies and growth rates from the data and to compare those values to the ones used to generate the data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">data</span> <span class="o">=</span> <span class="n">f1</span> <span class="o">+</span> <span class="n">f2</span> <span class="o">+</span> <span class="n">f3</span>
<span class="n">plot_as_line</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_dmd_intro_12_0.png" src="../_images/notebooks_dmd_intro_12_0.png" />
</div>
</div>
<p>Before we get to the DMD, let’s try a second approach to visualize the dataset. This time, we roll out the temporal evolution and show the full dataset as image. The function’s value is indicated by color (yellow for high and blue for low values). To make the analysis more challening, we generate a second dataset that is corrupted by noise. The noise is sampled from a Gaussian distribution with non-zero mean, which introduces some bias.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">show_as_image</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">X</span><span class="p">,</span> <span class="n">T</span><span class="p">):</span>
    <span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">pcolormesh</span><span class="p">(</span><span class="n">T</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">X</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">real</span><span class="p">,</span> <span class="n">shading</span><span class="o">=</span><span class="s2">&quot;nearest&quot;</span><span class="p">)</span>
    <span class="n">ax2</span><span class="o">.</span><span class="n">pcolormesh</span><span class="p">(</span><span class="n">T</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">X</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">imag</span><span class="p">,</span> <span class="n">shading</span><span class="o">=</span><span class="s2">&quot;nearest&quot;</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;$t$&quot;</span><span class="p">)</span>
    <span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;$t$&quot;</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;$x$&quot;</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;$Re(f)$&quot;</span><span class="p">)</span>
    <span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;$Im(f)$&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">show_as_image</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">X</span><span class="p">,</span> <span class="n">T</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_dmd_intro_15_0.png" src="../_images/notebooks_dmd_intro_15_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">data_noisy</span> <span class="o">=</span> <span class="n">data</span> <span class="o">+</span> <span class="n">pt</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">pt</span><span class="o">.</span><span class="n">ones_like</span><span class="p">(</span><span class="n">X</span><span class="p">),</span> <span class="mi">2</span><span class="o">*</span><span class="n">pt</span><span class="o">.</span><span class="n">ones_like</span><span class="p">(</span><span class="n">X</span><span class="p">))</span>
<span class="n">show_as_image</span><span class="p">(</span><span class="n">data_noisy</span><span class="p">,</span> <span class="n">X</span><span class="p">,</span> <span class="n">T</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_dmd_intro_16_0.png" src="../_images/notebooks_dmd_intro_16_0.png" />
</div>
</div>
</div>
<div class="section" id="Singular-Value-Decomposition-(SVD)">
<h3>Singular Value Decomposition (SVD)<a class="headerlink" href="#Singular-Value-Decomposition-(SVD)" title="Permalink to this headline">¶</a></h3>
<p>The first step of the DMD consists of computing the SVD of the <span class="math notranslate nohighlight">\(\mathbf{X}\)</span> matrix. Our dataset is already organized in the right way with the snapshots forming the columns of the data matrix. We compute the SVD on both the default and the noisy datasets. Note that instead of creating two data matrices <span class="math notranslate nohighlight">\(\mathbf{X}\)</span> and <span class="math notranslate nohighlight">\(\mathbf{X}^\prime\)</span>, we work with slicing; refer to <a class="reference internal" href="linear_algebra_basics.html"><span class="doc">this notebook</span></a> for more information on slicing tensors.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">U</span><span class="p">,</span> <span class="n">s</span><span class="p">,</span> <span class="n">VH</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">svd</span><span class="p">(</span><span class="n">data</span><span class="p">[:,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="n">Un</span><span class="p">,</span> <span class="n">sn</span><span class="p">,</span> <span class="n">VHn</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">svd</span><span class="p">(</span><span class="n">data_noisy</span><span class="p">[:,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p>In the original dataset, three non-zero singular values are identified. Adding noise yields many more singular values with smaller but non-zero values. Also the largest singular value differs slightly from that of the clean data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">size</span><span class="p">()[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">s</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;C0&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;data&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">sn</span><span class="o">.</span><span class="n">size</span><span class="p">()[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">sn</span><span class="p">,</span> <span class="n">ec</span><span class="o">=</span><span class="s2">&quot;C1&quot;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;noisy data&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;$i$&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;$\sigma_i$&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">30</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_dmd_intro_20_0.png" src="../_images/notebooks_dmd_intro_20_0.png" />
</div>
</div>
</div>
<div class="section" id="Truncating">
<h3>Truncating<a class="headerlink" href="#Truncating" title="Permalink to this headline">¶</a></h3>
<p>Truncation is an essential step of the DMD to reduce the computational effort but also to remove potential noise from the data. Here, we know the true rank of the data matrix, which is in general not the case. When working with real datasets, one can determine a suitable truncation by trial and error or by using some other criterion as demonstrated in the <a class="reference internal" href="svd_ht.html"><span class="doc">this example</span></a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">rank</span> <span class="o">=</span> <span class="mi">3</span>
<span class="n">Ur</span><span class="p">,</span> <span class="n">sr</span><span class="p">,</span> <span class="n">Vr</span> <span class="o">=</span> <span class="n">U</span><span class="p">[:,</span> <span class="p">:</span><span class="n">rank</span><span class="p">],</span> <span class="n">s</span><span class="p">[:</span><span class="n">rank</span><span class="p">],</span> <span class="n">VH</span><span class="o">.</span><span class="n">conj</span><span class="p">()</span><span class="o">.</span><span class="n">T</span><span class="p">[:,</span> <span class="p">:</span><span class="n">rank</span><span class="p">]</span>
<span class="n">Unr</span><span class="p">,</span> <span class="n">snr</span><span class="p">,</span> <span class="n">Vnr</span> <span class="o">=</span> <span class="n">Un</span><span class="p">[:,</span> <span class="p">:</span><span class="n">rank</span><span class="p">],</span> <span class="n">sn</span><span class="p">[:</span><span class="n">rank</span><span class="p">],</span> <span class="n">VHn</span><span class="o">.</span><span class="n">conj</span><span class="p">()</span><span class="o">.</span><span class="n">T</span><span class="p">[:,</span> <span class="p">:</span><span class="n">rank</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Eigen-decomposition">
<h3>Eigen decomposition<a class="headerlink" href="#Eigen-decomposition" title="Permalink to this headline">¶</a></h3>
<p>Next, we compute the reduced linear operator <span class="math notranslate nohighlight">\(\tilde{\mathbf{A}}\)</span> using the time-shifted matrix <span class="math notranslate nohighlight">\(\mathbf{X}^\prime\)</span> and the truncated SVD.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">At</span> <span class="o">=</span> <span class="n">Ur</span><span class="o">.</span><span class="n">conj</span><span class="p">()</span><span class="o">.</span><span class="n">T</span> <span class="o">@</span> <span class="n">data</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">:]</span> <span class="o">@</span> <span class="n">Vr</span> <span class="o">@</span> <span class="n">pt</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="mf">1.0</span><span class="o">/</span><span class="n">sr</span><span class="p">)</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">pt</span><span class="o">.</span><span class="n">cfloat</span><span class="p">)</span>
<span class="n">Atn</span> <span class="o">=</span> <span class="n">Unr</span><span class="o">.</span><span class="n">conj</span><span class="p">()</span><span class="o">.</span><span class="n">T</span> <span class="o">@</span> <span class="n">data</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">:]</span> <span class="o">@</span> <span class="n">Vnr</span> <span class="o">@</span> <span class="n">pt</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="mf">1.0</span><span class="o">/</span><span class="n">snr</span><span class="p">)</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">pt</span><span class="o">.</span><span class="n">cfloat</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The real and imaginary parts of the operator’s eigenvalues tell us about the time-dynamics of individual modes. First, we plot the eigenvalues in the imaginary plane together with a unit-circle. Note that for the time-discrete system, the modes are:</p>
<ul class="simple">
<li><p>stable if the eigenvalues are on the circle</p></li>
<li><p>growing (unstable) if the eigenvalues are outside the circle</p></li>
<li><p>shrinking (stable) if the eigenvalues are inside the circle</p></li>
</ul>
<p>As can be seen in the image below, the dynamics of all three components are identified correctly. The noise introduces some error. However, thanks to the truncation, we already removed a significant portion of the noise, and the identified dynamics remain close to the ground truth.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">val</span><span class="p">,</span> <span class="n">vec</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">eig</span><span class="p">(</span><span class="n">At</span><span class="p">)</span>
<span class="n">valn</span><span class="p">,</span> <span class="n">vecn</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">eig</span><span class="p">(</span><span class="n">Atn</span><span class="p">)</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">pi</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">pt</span><span class="o">.</span><span class="n">cos</span><span class="p">(</span><span class="n">p</span><span class="p">),</span> <span class="n">pt</span><span class="o">.</span><span class="n">sin</span><span class="p">(</span><span class="n">p</span><span class="p">),</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;k&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">pt</span><span class="o">.</span><span class="n">cos</span><span class="p">(</span><span class="n">p</span><span class="p">),</span><span class="mi">0</span><span class="p">,</span> <span class="n">pt</span><span class="o">.</span><span class="n">sin</span><span class="p">(</span><span class="n">p</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;k&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">val</span><span class="o">.</span><span class="n">real</span><span class="p">,</span> <span class="n">val</span><span class="o">.</span><span class="n">imag</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;data&quot;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">valn</span><span class="o">.</span><span class="n">real</span><span class="p">,</span> <span class="n">valn</span><span class="o">.</span><span class="n">imag</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;noisy data&quot;</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">ev</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">val</span><span class="p">):</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">real</span><span class="o">+</span><span class="mf">0.02</span><span class="p">,</span> <span class="n">ev</span><span class="o">.</span><span class="n">imag</span><span class="o">+</span><span class="mf">0.02</span><span class="p">,</span> <span class="sa">r</span><span class="s2">&quot;$\lambda_</span><span class="si">{:d}</span><span class="s2">$&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="s2">&quot;shrinking modes&quot;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">,</span> <span class="s2">&quot;growing modes&quot;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s2">&quot;equal&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;$Re(\lambda_i)$&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;$Im(\lambda_i)$&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_dmd_intro_26_0.png" src="../_images/notebooks_dmd_intro_26_0.png" />
</div>
</div>
<p>Let’s now compute the eigenvalues of the corresponding continuous linear operator. From the real and imaginary parts of all <span class="math notranslate nohighlight">\(\tilde{\lambda}_i\)</span>, we can directly read off growth rates and frequencies.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">dt</span> <span class="o">=</span> <span class="n">t</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">lambda_t</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">val</span><span class="p">)</span> <span class="o">/</span> <span class="n">dt</span>
<span class="n">lambdan_t</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">valn</span><span class="p">)</span> <span class="o">/</span> <span class="n">dt</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;clean dataset&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">lambda_t</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;Mode </span><span class="si">{:d}</span><span class="s2"> has a growth rate of </span><span class="si">{:+2.3f}</span><span class="s2"> and oscillates with a frequency of </span><span class="si">{:2.3f}</span><span class="s2">/(2π)Hz.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">l</span><span class="o">.</span><span class="n">real</span><span class="p">,</span> <span class="n">l</span><span class="o">.</span><span class="n">imag</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">noisy dataset&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">lambdan_t</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;Mode </span><span class="si">{:d}</span><span class="s2"> has a growth rate of </span><span class="si">{:+2.3f}</span><span class="s2"> and oscillates with a frequency of </span><span class="si">{:2.3f}</span><span class="s2">/(2π)Hz.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">l</span><span class="o">.</span><span class="n">real</span><span class="p">,</span> <span class="n">l</span><span class="o">.</span><span class="n">imag</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
clean dataset
Mode 0 has a growth rate of -0.050 and oscillates with a frequency of 2.300/(2π)Hz.
Mode 1 has a growth rate of +0.100 and oscillates with a frequency of 2.800/(2π)Hz.
Mode 2 has a growth rate of -0.000 and oscillates with a frequency of 0.600/(2π)Hz.

noisy dataset
Mode 0 has a growth rate of -0.134 and oscillates with a frequency of 2.315/(2π)Hz.
Mode 1 has a growth rate of +0.075 and oscillates with a frequency of 2.796/(2π)Hz.
Mode 2 has a growth rate of -0.002 and oscillates with a frequency of 0.605/(2π)Hz.
</pre></div></div>
</div>
<p>With the clean dataset as input, DMD perfectly identifies growth rates and frequencies of the underlying functions used to create the dataset. The noise corrupts the values slightly, but the results remain meaningful, too.</p>
</div>
<div class="section" id="DMD-modes">
<h3>DMD modes<a class="headerlink" href="#DMD-modes" title="Permalink to this headline">¶</a></h3>
<p>In the final part of this notebook, we reconstruct the eigenvectors of the full linear operator, the so-called DMD modes, and visualize the difference between projected and exact DMD modes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">phi</span> <span class="o">=</span> <span class="n">data</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">:]</span> <span class="o">@</span> <span class="n">Vr</span> <span class="o">@</span> <span class="n">pt</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="mf">1.0</span><span class="o">/</span><span class="n">sr</span><span class="p">)</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">pt</span><span class="o">.</span><span class="n">cfloat</span><span class="p">)</span> <span class="o">@</span> <span class="n">vec</span>
<span class="n">phin</span> <span class="o">=</span> <span class="n">data_noisy</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">:]</span> <span class="o">@</span> <span class="n">Vnr</span> <span class="o">@</span> <span class="n">pt</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="mf">1.0</span><span class="o">/</span><span class="n">snr</span><span class="p">)</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">pt</span><span class="o">.</span><span class="n">cfloat</span><span class="p">)</span> <span class="o">@</span> <span class="n">vecn</span>
<span class="n">phi_pro</span> <span class="o">=</span> <span class="n">Ur</span> <span class="o">@</span> <span class="n">vec</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;C</span><span class="si">{:1d}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">)]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">rank</span><span class="p">):</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">phi</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">real</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;$\varphi_</span><span class="si">{:d}</span><span class="s2">$&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">phin</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">real</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;$\varphi_</span><span class="si">{:d}</span><span class="s2">$, noisy&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">phi_pro</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">real</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;-.&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;$\tilde{\varphi}&quot;</span> <span class="o">+</span> <span class="sa">r</span><span class="s2">&quot;_</span><span class="si">{:d}</span><span class="s2">$&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_dmd_intro_31_0.png" src="../_images/notebooks_dmd_intro_31_0.png" />
</div>
</div>
<p>Each mode characterizes the shape of the corresponding input function very well. The noise is still visible but the reconstructed modes are close to the reference. The projected DMD modes look qualitatively similar to their exact counterparts but are not identical for the present dataset.</p>
<p>The first version of this notebook was created by <strong>Jannis Anderson</strong> during his time as student assisant at TU Braunschweig’s <a class="reference external" href="https://www.tu-braunschweig.de/ism">Institute of Fluid Mechanics</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="svd_cylinder.html" class="btn btn-neutral float-right" title="SVD of the flow past a circular cylinder" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="ipsp_explorer.html" class="btn btn-neutral float-left" title="PSP data explorer and loader" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, flowTorch contributors.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>